### ISSUE TITLE
Scikit-learn example for random forest with Joblib

### ISSUE THREAD

**Issue by mrocklin** [dask/dask-ml#231]
It would be nice to see an example using Scikit-Learn's RandomForest parallelized with Dask using the joblib/dask connection.  This is a common question, and so probably has high value.

My understanding is that this will depend on the 0.20 release of scikit-learn first, though I could be mistaken about this.

It would also be useful for this to be smoothly runnable on [dask-examples](https://github.com/dask/dask-examples#contributing)

**Comment by TomAugspurger** [comment]
> My understanding is that this will depend on the 0.20 release of scikit-learn first, though I could be mistaken about this.

Yes, that's correct.

**Comment by mrocklin** [comment]
Are there common enhancements beyond just RandomForest?  For example does it make sense to do RandomForest within a GridSearchCV?

**Comment by TomAugspurger** [comment]
> does it make sense to do RandomForest within a GridSearchCV?

Yep, certainly. That will introduce two levels of parallelism though, which may not be well-suited for dask-examples.

There are other ensemble methods, like http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingClassifier.html#sklearn.ensemble.GradientBoostingClassifier, that aren't parallelized at the estimator level.

**Comment by js3711** [comment]
Does this need sklearn v0.20 to work?  For some reason I thought there was (maybe used to be) some working examples in the documentation for grid-search

**Comment by TomAugspurger** [comment]
RandomForest (and a few others) currently hardcode their joblib backends, so they would ignore the `parallel_backend` context.  https://github.com/scikit-learn/scikit-learn/pull/9486, which should be in 0.20, includes the changes to remove the hardcoding.

**Comment by js3711** [comment]
It looks like there's already an example of something similar in the dask-ml documentation:
http://dask-ml.readthedocs.io/en/latest/examples/joblib-distributed.html
Should this be tagged as dependent on scitkit learn v0.20.0

Also, I took a quick look at sklearn dev branch.  It looks like they are still using threading:
https://github.com/scikit-learn/scikit-learn/blob/f158e2dfe2af1b23ae3f9d86c598013b2c155c3f/sklearn/ensemble/forest.py#L329
a call to parallel_backend now override this in the new joblib?

**Comment by TomAugspurger** [comment]
To clarify, the dask.distributed backend works fine with scikit-learn 0.19
for most estimators.
http://dask-ml.readthedocs.io/en/latest/examples/joblib-distributed.html
should probably be updated to use something like GridSearchCV, similar to
http://dask-ml.readthedocs.io/en/latest/joblib.html

The PR https://github.com/scikit-learn/scikit-learn/pull/9486 isn't merged
yet. Once that's in, it'll use the new joblib API for prefers / requires
shared memory.

**Comment by js3711** [comment]
Got it.  Thanks Tom, so the dask.distributed backend should work for sklearn estimators that do not explicitly mention the backend argument in Parallel.

I was wondering why RandomForest did not appear to light up the dask scheduler :)

**Comment by ogrisel** [comment]
FYI, I just merged 0.12.2 into the master branch of scikit-learn.
